crates/ruff_python_ast/ast.toml

```
@@ -80,39 +80,367 @@ add_suffix_to_is_methods = true
 anynode_is_label = "expression"
 rustdoc = "/// See also [expr](https://docs.python.org/3/library/ast.html#ast.expr)"
 
-[Expr.nodes]
-ExprBoolOp = {}
-ExprNamed = {}
-ExprBinOp = {}
-ExprUnaryOp = {}
-ExprLambda = {}
-ExprIf = {}
-ExprDict = {}
-ExprSet = {}
-ExprListComp = {}
-ExprSetComp = {}
-ExprDictComp = {}
-ExprGenerator = {}
-ExprAwait = {}
-ExprYield = {}
-ExprYieldFrom = {}
-ExprCompare = {}
-ExprCall = {}
-ExprFString = {}
-ExprStringLiteral = {}
-ExprBytesLiteral = {}
-ExprNumberLiteral = {}
-ExprBooleanLiteral = {}
-ExprNoneLiteral = {}
-ExprEllipsisLiteral = {}
-ExprAttribute = {}
-ExprSubscript = {}
-ExprStarred = {}
-ExprName = {}
-ExprList = {}
-ExprTuple = {}
-ExprSlice = {}
-ExprIpyEscapeCommand = {}
+[Expr.nodes.ExprBoolOp]
+rustdoc = "/// See also [BoolOp](https://docs.python.org/3/library/ast.html#ast.BoolOp)"
+
+[[Expr.nodes.ExprBoolOp.fields]]
+name = "op"
+type = "BoolOp"
```

Should we use inline arrays? I find the syntax very verbose to read and find it hard to get a sense of a nodes fields
```suggestion
fields = [
	{ name = "op", type = "BoolOp" },
	{ name = "values", type = "Expr", seq = true },
	...
}
```

The downside of this is that inline tables can't be multiline (this can be a problem with comments)

---

In fairness to @Glyphack I had suggested the syntax he's using here https://github.com/astral-sh/ruff/pull/16285#discussion_r1969927355!

Another possibility, since `name` is required and unique, would be

```toml
[Expr.nodes.ExprBoolOp]
fields = {
  op = { type = "BoolOp" },
  values = { type = "Expr", seq = true }
}
```

---

Oh, I didn't see that one.

The second doesn't work, I think, because toml inline tables don't allow line breaks

---

yes unfortunately it's not possible to have line break inside an inline table.
To make that work we need to have a table `[Expr.nodes.ExprBoolOp.fields.op]` which takes it more verbose I think?

But I think instead of the inline table using an inline array would solve the problem of having all the fields in one section and we only need to write an extra "name = ..." does sound good to me.

But one thing to keep in mind is that if we use inline table for the field attributes and have tables inside attributes(for example if we apply this comment https://github.com/astral-sh/ruff/pull/16285#discussion_r1971093809) it might get very messy example:

```
[Expr.nodes.ExprBoolOp]
fields = [
  { name = "values", type = { kind = "sequence", element-type = "Expr" } }
]
```

I feel we end up making type a more complex field in the end.

---

That's a fair concern. Maybe we need ungrammar after all (just joking). 

The alternative is to change `type` to accept some form of DSL (which is what ungrammar would do as well)

* `Expr` (single expression)
* `Expr?` (optional expression)
* `Expr*` or `[Expr]` (Vec)
* `&Expr*` or `&[Expr]` (Slice) maybe there's a better way of spelling this

---

Let me try this out 👍. I'll use `Expr*` and `&Expr*` syntax for now. 

Also if we decide to make this a DSL we can just drop the name since it can be easily added by having something like:
`value:Expr` and eliminates more of the structure we have now.

```
[Expr.nodes.ExprBoolOp]
fields = [
  "op:BoolOp",
  "values:Expr*"
]
```

---

I don't mind the inline table. I probably prefer it because it's less *magic*.

---

I implemented this. A question, do you think we would want to nest these rules? for example creating a vector with optional values. The rule for it would be `Expr?*`. It doesn't look good so I intentionally left this out of the current implementation. We can introduce parenthesis if we want to do it.
This was the only question, the code generation logic also looks better now.
Let me know what do you think about the implementation.

https://github.com/astral-sh/ruff/pull/16285/files#diff-c52a3d2dd4a45e516b700b71c4d38350f80c2796c8eb236f05e7a2dddf3689f1R145

---

A question, do you think we would want to nest these rules? for example creating a vector with optional values. The rule for it would be `Expr?*`. It doesn't look good so I intentionally left this out of the current implementation. We can introduce parenthesis if we want to do it.

I think the correct way to do this also would be to not only have `FieldType` class but have a class for each data type it represents for example `FieldTypeSeq` for fields that are sequence:

```python
class FieldTypeSeq:
  element_ty: FieldType
  is_slice: bool


class FieldTypeOptional:
  some_ty: FieldType
```

I did not want to make code more complex without a reason right now, it's probably needed when porting more code to generate script.

Let me know what do you think about the implementation.

https://github.com/astral-sh/ruff/pull/16285/files#diff-c52a3d2dd4a45e516b700b71c4d38350f80c2796c8eb236f05e7a2dddf3689f1R145

---

Do we have use cases today where we have optional sequence elements (I hope not). I think it's very unlikely that we'll need this in the future because we want a node for each element in a sequence to at least preserve the node's range (each element in a sequence should implement the `AstNode` trait and `None` can't do that)

---

Just checked and there's no use of this (I searched for Vec<Option in the codebase)